File: backend/app/ws/frames.py
"""Binary media frame codec for ``/ws/live``.

Implements the fixed little-endian header defined in ``PROTOCOL.md`` section 2::

     offset  size  field
       0      1    tag   (uint8)   — stream type
       1      8    ts    (uint64)  — capture/emit time, ms since epoch (LE)
       9      ..   payload         — raw bytes (PCM or JPEG)

WebSocket frames are already length-delimited, so no length prefix is required.
"""

from __future__ import annotations

import struct
import time
from enum import IntEnum

#: Size of the fixed binary header in bytes (1-byte tag + 8-byte LE uint64 ts).
HEADER_SIZE: int = 9

#: ``struct`` format for the header: little-endian uint8 + uint64.
_HEADER_STRUCT = struct.Struct("<BQ")


class FrameTag(IntEnum):
    """Stream type tags for binary frames (see ``PROTOCOL.md``)."""

    INPUT_AUDIO = 0x01  # client -> server, PCM16 LE 16 kHz mono
    INPUT_VIDEO = 0x02  # client -> server, JPEG single frame
    OUTPUT_AUDIO = 0x03  # server -> client, PCM16 LE 24 kHz mono


def now_ms() -> int:
    """Return the current wall-clock time in milliseconds since the epoch."""
    return int(time.time() * 1000)


def encode_frame(tag: int | FrameTag, payload: bytes, ts: int | None = None) -> bytes:
    """Encode a media payload into a wire frame.

    Args:
        tag: The :class:`FrameTag` (or raw ``uint8``) identifying the stream.
        payload: Raw media bytes (PCM samples or a JPEG frame).
        ts: Timestamp in ms since epoch. Defaults to :func:`now_ms`.

    Returns:
        The header (9 bytes) followed by ``payload``.

    Raises:
        ValueError: If ``tag`` or ``ts`` does not fit the header field widths.
    """
    tag_int = int(tag)
    if not 0 <= tag_int <= 0xFF:
        raise ValueError(f"tag must fit in a uint8, got {tag_int}")
    timestamp = now_ms() if ts is None else int(ts)
    if not 0 <= timestamp <= 0xFFFFFFFFFFFFFFFF:
        raise ValueError(f"ts must fit in a uint64, got {timestamp}")
    return _HEADER_STRUCT.pack(tag_int, timestamp) + bytes(payload)


def decode_frame(data: bytes) -> tuple[int, int, bytes]:
    """Decode a wire frame into ``(tag, ts, payload)``.

    Args:
        data: A full binary frame as received from the socket.

    Returns:
        A tuple of ``(tag, ts_ms, payload_bytes)``. ``tag`` is returned as a
        plain ``int`` so unknown/forward-compatible tags do not raise; callers
        may coerce to :class:`FrameTag` when they expect a known value.

    Raises:
        ValueError: If ``data`` is shorter than the fixed header.
    """
    if len(data) < HEADER_SIZE:
        raise ValueError(
            f"frame too short: need >= {HEADER_SIZE} bytes, got {len(data)}"
        )
    tag, ts = _HEADER_STRUCT.unpack_from(data, 0)
    payload = bytes(data[HEADER_SIZE:])
    return tag, ts, payload

File: mobile/lib/protocol/frames.dart
import 'dart:typed_data';

import 'protocol.dart';

/// Codec for the fixed binary media-frame header defined in `PROTOCOL.md` §2.
///
/// Wire layout (little-endian, no length field — WebSocket frames are already
/// length-delimited):
///
/// ```
///  offset  size  field
///  ------  ----  ----------------------------------------------
///    0      1    tag   (uint8)   — stream type (see [FrameTag])
///    1      8    ts    (uint64)  — capture/emit time, ms since epoch (LE)
///    9      ..   payload         — raw bytes (PCM or JPEG)
/// ```
///
/// This is the most contract-sensitive code in the app: it must produce and
/// consume bytes identically to the Python backend. Round-trip and
/// known-vector tests live in `test/frames_test.dart`.
abstract final class MediaFrame {
  /// Size of the fixed header in bytes (`uint8` tag + `uint64` timestamp).
  static const int headerSize = 9;

  /// Offset of the timestamp field within the header.
  static const int _tsOffset = 1;

  /// Encode a media frame into a single contiguous [Uint8List].
  ///
  /// [tag] is one of [FrameTag.inputAudio] / [FrameTag.inputVideo] /
  /// [FrameTag.outputAudio]. [timestampMs] is milliseconds since the Unix epoch
  /// and is written as an unsigned 64-bit little-endian integer. [payload] is
  /// the raw media bytes (PCM16 or JPEG) appended verbatim.
  ///
  /// Throws [ArgumentError] if [tag] does not fit in a single byte or
  /// [timestampMs] is negative.
  static Uint8List encode({
    required int tag,
    required int timestampMs,
    required Uint8List payload,
  }) {
    if (tag < 0 || tag > 0xFF) {
      throw ArgumentError.value(tag, 'tag', 'must be a single byte (0–255)');
    }
    if (timestampMs < 0) {
      throw ArgumentError.value(
        timestampMs,
        'timestampMs',
        'must be non-negative',
      );
    }

    final out = Uint8List(headerSize + payload.length);
    final header = ByteData.sublistView(out, 0, headerSize);
    header.setUint8(0, tag);
    // setUint64 with Endian.little writes the 8-byte LE timestamp at offset 1.
    header.setUint64(_tsOffset, timestampMs, Endian.little);
    out.setRange(headerSize, out.length, payload);
    return out;
  }

  /// Convenience wrapper that stamps the frame with the current wall-clock time.
  static Uint8List encodeNow({
    required int tag,
    required Uint8List payload,
  }) =>
      encode(
        tag: tag,
        timestampMs: DateTime.now().millisecondsSinceEpoch,
        payload: payload,
      );

  /// Decode a media frame received from the wire.
  ///
  /// Accepts any [List<int>] (the `web_socket_channel` binary payload type) and
  /// returns a [DecodedFrame]. The returned [DecodedFrame.payload] is a zero-copy
  /// view over [bytes] when possible, so do not mutate [bytes] afterwards.
  ///
  /// Throws [FormatException] if the buffer is shorter than [headerSize].
  static DecodedFrame decode(List<int> bytes) {
    final data = bytes is Uint8List ? bytes : Uint8List.fromList(bytes);
    if (data.length < headerSize) {
      throw FormatException(
        'media frame too short: ${data.length} bytes '
        '(need at least $headerSize)',
      );
    }
    final header = ByteData.sublistView(data, 0, headerSize);
    final tag = header.getUint8(0);
    final ts = header.getUint64(_tsOffset, Endian.little);
    final payload = Uint8List.sublistView(data, headerSize);
    return DecodedFrame(tag: tag, timestampMs: ts, payload: payload);
  }
}

/// A decoded binary media frame: its [tag], [timestampMs], and [payload].
class DecodedFrame {
  const DecodedFrame({
    required this.tag,
    required this.timestampMs,
    required this.payload,
  });

  /// Stream type — one of the [FrameTag] constants.
  final int tag;

  /// Capture/emit time in milliseconds since the Unix epoch.
  final int timestampMs;

  /// Raw media bytes (PCM16 or JPEG), a view over the source buffer.
  final Uint8List payload;

  @override
  String toString() =>
      'DecodedFrame(tag: 0x${tag.toRadixString(16).padLeft(2, '0')}, '
      'ts: $timestampMs, payload: ${payload.length}B)';
}
